### DISCUSSION TITLE
Prisma.upsert() should support the value of null in where condition with unique key

### DISCUSSION THREAD

**Discussion by MarieMinasyan** [prisma/prisma#22897] [Q&A]
### Question

Hello,
I have a use case with a unique constraint on 3 columns, one of which is a nullable foreign key.
With the unique constraint, I still have only one record where the first two columns have specific values and the third one is null.
I'd like to be able to upsert into my table with:

```
where: {
  firstColumn_secondColumn_thirdColumn: {
    firstColumn: "value",
    secondColumn: "value",
    thirdColumn: null,
  }
}
```
Is there a way to do this?
Thank you.

### Information about Prisma Schema, Client Queries and Environment (optional)

- OS: Ubuntu
- Database: PostgreSQL
- Node.js version: v18.17.1

**Comment by Jolg42** [comment]
Bonjour Marie 👋🏼 

One quick question, did you try to do the upsert with Prisma Client? What happened?
Also, which version of Prisma Client are you using here?

**Reply by MarieMinasyan** [reply]
Hi,
Yes I tried with Prisma client, I get the following `PrismaClientValidationError`, `Argument "gameId" must not be null.`
My Prisma Client version is 5.2.0

**Comment by ludralph** [comment]
Hi @MarieMinasyan 👋 

Prisma does not support the use of `null` in the where condition of the `upsert()` method when dealing with unique constraints that involve nullable fields.

As mentioned in this closed [issue](https://github.com/prisma/prisma/issues/6359#issuecomment-825649937), this is by design. Prisma Client requires a unique condition to pin exactly one record in the database for the `upsert()` operation. This is because `upsert()` is designed to operate on a single record, either updating it if it exists or creating it if it does not. Using `null` in a unique constraint can lead to ambiguity because, in SQL, NULL is not considered equal to another NULL value, which means that the unique constraint does not behave as expected when null values are involved.

As a workaround, you may need to use raw queries or 2-3 queries in your application as mentioned [here](https://github.com/prisma/prisma/issues/6359#issuecomment-819605011).

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by MarieMinasyan** [reply]
We use the [UNIQUE NULLS NOT DISTINCT](https://www.postgresql.org/docs/current/ddl-constraints.html#DDL-CONSTRAINTS-UNIQUE-CONSTRAINTS) constraint to make sure that null value is handled as unique in the Database.
Thank you for your answers.
